go/arrow/util/protobuf_reflect.go

```diff
@@ -0,0 +1,762 @@
+		protoreflect.Int64Kind:    arrow.PrimitiveTypes.Int64,
+		protoreflect.Sint32Kind:   arrow.PrimitiveTypes.Int32,
+		protoreflect.Sint64Kind:   arrow.PrimitiveTypes.Int64,
+		protoreflect.Uint32Kind:   arrow.PrimitiveTypes.Uint32,
+		protoreflect.Uint64Kind:   arrow.PrimitiveTypes.Uint64,
+		protoreflect.Fixed32Kind:  arrow.PrimitiveTypes.Uint32,
+		protoreflect.Fixed64Kind:  arrow.PrimitiveTypes.Uint64,
+		protoreflect.Sfixed32Kind: arrow.PrimitiveTypes.Int32,
+		protoreflect.Sfixed64Kind: arrow.PrimitiveTypes.Int64,
+		protoreflect.FloatKind:    arrow.PrimitiveTypes.Float32,
+		protoreflect.DoubleKind:   arrow.PrimitiveTypes.Float64,
+		//Binary
+		protoreflect.StringKind: arrow.BinaryTypes.String,
+		protoreflect.BytesKind:  arrow.BinaryTypes.Binary,
+		//Fixed Width
+		protoreflect.BoolKind: arrow.FixedWidthTypes.Boolean,
+		// Special
+		protoreflect.EnumKind:    nil,
+		protoreflect.MessageKind: nil,
+	}
+	dt = typeMap[pfr.descriptor.Kind()]
+
+	switch pfr.arrowType() {
+	case arrow.DENSE_UNION:
+		dt = pfr.asUnion().getDataType()
+	case arrow.DICTIONARY:
+		dt = pfr.asDictionary().getDataType()
+	case arrow.LIST:
+		dt = pfr.asList().getDataType()
+	case arrow.MAP:
+		dt = pfr.asMap().getDataType()
+	case arrow.STRUCT:
+		dt = pfr.asStruct().getDataType()
+	}
+
+	return dt
+}
+
+func (sf SuperField) AppendValueOrNull(b array.Builder, mem memory.Allocator) {
+	pv := sf.protoreflectValue()
+	fd := sf.getDescriptor()
+
+	if sf.isNull() {
+		b.AppendNull()
+		return
+	}
+
+	switch b.Type().ID() {
+	case arrow.STRING:
+		b.(*array.StringBuilder).Append(pv.String())
+	case arrow.BINARY:
+		b.(*array.BinaryBuilder).Append(pv.Bytes())
+	case arrow.INT32:
+		b.(*array.Int32Builder).Append(int32(pv.Int()))
+	case arrow.INT64:
+		b.(*array.Int64Builder).Append(pv.Int())
+	case arrow.FLOAT64:
+		b.(*array.Float64Builder).Append(pv.Float())
+	case arrow.UINT32:
+		b.(*array.Uint32Builder).Append(uint32(pv.Uint()))
+	case arrow.UINT64:
+		b.(*array.Uint64Builder).Append(pv.Uint())
+	case arrow.BOOL:
+		b.(*array.BooleanBuilder).Append(pv.Bool())
+	case arrow.DENSE_UNION:
+		ub := b.(array.UnionBuilder)
+		pur := sf.asUnion()
+		if pur.whichOne() == -1 {
+			ub.AppendNull()
+			break
+		}
+		ub.Append(pur.whichOne())
+		cb := ub.Child(int(pur.whichOne()))
+		SuperField{
+			parent:             sf.parent,
+			protobufReflection: pur.getField(),
+			Field:              pur.arrowField(),
+		}.AppendValueOrNull(cb, mem)
+	case arrow.DICTIONARY:
+		db := b.(array.DictionaryBuilder)
+		err := db.AppendValueFromString(string(fd.Enum().Values().ByNumber(pv.Enum()).Name()))
+		if err != nil {
+			fmt.Println(err)
+		}
+	case arrow.STRUCT:
+		sb := b.(*array.StructBuilder)
+		sb.Append(true)
+		for i, field := range sf.Field.Type.(*arrow.StructType).Fields() {
+			SuperField{
+				parent:             sf.parent,
+				protobufReflection: sf.asStruct().getFieldByName(field.Name),
+				Field:              field,
+			}.AppendValueOrNull(sb.FieldBuilder(i), mem)
+		}
```

Comment:
can you extract the `SuperField{}` creation from the loop, and just use the same instance over and over, just updating the members?

Reply:
that's a nice idea